graphql/schemabuilder/function.go

```diff
@@ -111,6 +118,62 @@ func (sb *schemaBuilder) buildFederatedFunction(typ reflect.Type, m *method) (*g
 	return field, nil
 }
 
+// buildShadowObjectFederationFunction builds a federation object and a field func that takes the
+// federation keys as args and constructs the shadow object. This is used for federated subqueries.
+// {
+//    __federation {
+//     [ObjectName]-[Service] (keys: Keys) {
+//       subQuery
+// 		}
+//   }
+// }
+func (sb *schemaBuilder) buildShadowObjectFederationFunction(typ reflect.Type, m *method) (*graphql.Field, error) {
+	funcCtx := &funcContext{typ: typ}
+
+	// Input type in the format of struct{keys: *ShadowObjectType}
+	// which are the federated keys pass into the field
+	shadowType := reflect.TypeOf(m.ShadowObjectType)
+	input := reflect.StructOf([]reflect.StructField{
+		{
+			Name: "Keys",
+			Type: reflect.SliceOf(shadowType),
+		},
+	})
+	in := make([]reflect.Type, 1)
+	in[0] = input
+
+	argParser, argType, _, err := funcCtx.getArgParserAndTyp(sb, in)
+	if err != nil {
+		return nil, oops.Wrapf(err, "Error parsing args for shadow object field")
+	}
+	funcCtx.hasArgs = argParser != nil
+	args, err := funcCtx.argsTypeMap(argType)
+	if err != nil {
+		return nil, err
+	}
+
+	// Return type is a nonnullable list of the shadow object type
+	returnType, err := sb.getType(reflect.TypeOf(m.ShadowObjectType))
+	if err != nil {
+		return nil, oops.Wrapf(err, "Invalid return type")
+	}
+	rType := &graphql.NonNull{Type: &graphql.List{Type: returnType}}
+
+	field := &graphql.Field{
+		Resolve: func(ctx context.Context, source, funcRawArgs interface{}, selectionSet *graphql.SelectionSet) (interface{}, error) {
+			args := funcRawArgs.(map[string]graphql.Type)
+			return args["keys"], nil
```

Comment:
Just curious, the funcRawRags has a non pointer type and our return type is a pointer, is that okay?

Reply:
Yup I think so since the funcRawRags are nonnullable

Reply:
Ah sorry what i meant is our arg looks like this 
struct {
    keys: []Device
}
whereas the `rType` is `[]*Device`.
Is that okay or we need to do some further translation?

Reply:
Ah, its working right now since the return type isn't expecting a pointer. We can change it so that they both use a pointer

Reply:
Does that matter? 
I think executor is smart enough to hand both cases?

Reply:
But it should be more efficient if we can return a pointer?

Reply:
How about we change both input and output to be a pointer?

Reply:
Is there any rule that the input keys could not be pointers?

Reply:
Ah, actually I misread the code, both the input type and return type have a pointer to the device!